imap_processing/ultra/l0/decom_ultra.py

```diff
@@ -1,211 +1,139 @@
-    m = value & 0xFFF  # Extract the 12 least significant bits for the mantissa
-
-    if e == 0:
-        return m
-    else:
-        return (4096 + m) << (e - 1)
-
-
-def decompress_binary(binary: str, width_bit: int, block: int) -> list:
-    """Decompress a binary string.
+    parser_helper = ParserHelper()
 
-    Decompress a binary string based on block-width encoding and
-    logarithmic compression.
+    for key, item in packet.data.items():
+        # Initialize the list for the first time
+        if key not in decom_data:
+            decom_data[key] = []
+        if key != decompressed_key:
+            decom_data[key].append(item.derived_value)
+        else:
+            decom_data[key].append(decompressed_data)
 
-    This function interprets a binary string where the first 'width_bit' bits
-    specifies the width of the following values. Each value is then extracted and
-    subjected to logarithmic decompression.
+    ccsds_data = CcsdsData(packet.header)
+    parser_helper.append_ccsds_fields(decom_data, ccsds_data)
 
-    Parameters
-    ----------
-    binary : str
-        A binary string containing the compressed data.
-    width_bit : int
-        The bit width that describes the width of data in the block
-    block : int
-        Number of values in each block
-
-    Returns
-    -------
-    list
-        A list of decompressed values.
 
-    Note: Equations from Section 1.2.1.1 Data Compression and Decompression Algorithms
-    in Ultra_algorithm_doc_rev2.pdf.
+def decom_ultra_apids(packet_file: str, xtce: str, test_apid: Optional[int] = None):
     """
-    current_position = 0
-    decompressed_values = []
-
-    while current_position < len(binary):
-        # Read the width of the block
-        width, current_position = read_n_bits(binary, width_bit, current_position)
-        # If width is 0 or None, we don't have enough bits left
-        if (
-            width is None
-            or len(decompressed_values) >= UltraParams.ULTRA_IMG_RATES.value.len_array
-        ):
-            print("hi")
-            break
-
-        # For each block, read 16 values of the given width
-        for _ in range(block):
-            # Ensure there are enough bits left to read the width
-            if len(binary) - current_position < width:
-                break
-
-            value, current_position = read_n_bits(binary, width, current_position)
-
-            # Log decompression
-            decompressed_values.append(log_decompression(value))
-
-    return decompressed_values
-
-
-def decom_ultra_packets(packet_file: str, xtce: str):
-    """
-    Unpack and decode ultra packets using CCSDS format and XTCE packet definitions.
+    Unpack and decode Ultra packets using CCSDS format and XTCE packet definitions.
 
     Parameters
     ----------
     packet_file : str
         Path to the CCSDS data packet file.
     xtce : str
         Path to the XTCE packet definition file.
+    test_apid : Optional[int]
```

Comment:
When you say "test", do you mean this will be removed, or just that you want a single apid? I'd remove the test qualifier if you plan on keeping it.
```suggestion
    apid : Optional[int]
```